### PR TITLE
Add examples to fmt precision

### LINKED ISSUE #24656: `std::fmt` needs doc for `precision := '*'` case
I spent a while looking at http://doc.rust-lang.org/nightly/std/fmt/index.html#syntax and comparing it against the rest of the API documentation there, but I saw no discussion of the meaning of the asterisk precision, e.g. what the format string `"{:.*}"` denotes.

The error message you get when you try to use it incorrectly is pretty inscrutable:
```rust
    println!("{:.*}", 0.0);
```
yields
```
<std macros>:3:11: 3:36 error: invalid reference to argument `1` (there is 1 argument)
<std macros>:3 print ! ( concat ! ( $ fmt , "\n" ) , $ ( $ arg ) * ) ) ;
                         ^~~~~~~~~~~~~~~~~~~~~~~~~
note: in expansion of format_args!
```

----

Manual experimentation seems to indicate that the meaning of `{:.*}` is that the next *two* parameters are consumed, where the first represents the fractional precision and the second the value to be printed.

As a concrete example, the following four statements all print the same thing:
```rust
fn main() {
    println!("Hello {1} is {2:.0$}", 5, "x", 0.01);
    println!("Hello {0} is {2:.1$}", "x", 5, 0.01);
    println!("Hello {} is {:.*}",    "x", 5, 0.01);
    println!("Hello {} is {2:.*}",   "x", 5, 0.01);
}
```
namely:
```
Hello x is 0.01000
Hello x is 0.01000
Hello x is 0.01000
Hello x is 0.01000
```

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #24656 

r? @pnkfelix 

I just added the examples, but if the wording needs expanded too, let me know what you think should be added :)

### CODE REVIEW

File: src/libcollections/fmt.rs [hunk 1]
```diff
@@ -407,6 +407,21 @@
 //! For floating-point types, this indicates how many digits after the decimal
 //! point should be printed.
 //!
+//! By example, these:
+//!
+//! ```
+//! println!("Hello {1} is {2:.0$}", 5, "x", 0.01);
+//! println!("Hello {0} is {2:.1$}", "x", 5, 0.01);
+//! println!("Hello {} is {:.*}",    "x", 5, 0.01);
+//! println!("Hello {} is {2:.*}",   "x", 5, 0.01);
```
Comment (Manishearth) [inline src/libcollections/fmt.rs]: Perhaps have a comment above each one reading out what each one means? Something like "Hello {first argument (x)} is {second argument (0.01) with precision specified in zeroeth argument (5)}"

File: src/libcollections/fmt.rs [hunk 2]
```diff
@@ -407,6 +407,21 @@
 //! For floating-point types, this indicates how many digits after the decimal
 //! point should be printed.
 //!
+//! By example, these:
```
Comment (Manishearth) [inline src/libcollections/fmt.rs]: Nit: s/By/For ?

**Comment by Manishearth** [comment]
r=me with or without comments addressed, unless you specifically wanted Felix's r :smile:

**Comment by pnkfelix** [comment]
@steveklabnik the additions seem fine; the suggestions noted by @Manishearth also seem like good changes.

Its funny (or perhaps sad); my original issue #24656 was that I could not find documentation for the actual meaning of `.*`.
 * Yet such documentation can clearly be seen here: https://github.com/rust-lang/rust/blob/master/src/libcollections/fmt.rs#L106
 * I have no explanation for how I overlooked that text in my repeated scans of the docs.  Maybe I was looking at the code directly in a checkout that predated your addition from PR #23493 ; anyway, sorry if my original issue text seemed shrill.

**Comment by pnkfelix** [comment]
Although, reading over that text from  https://github.com/rust-lang/rust/blob/master/src/libcollections/fmt.rs#L106 now, it seems like there is a *little* room for improvement.

Namely: While that text does describe the meaning of `{:.*}`, it does *not* describe the meaning of `{:.<num>}`.  I.e. I might have guessed that `{:.5}` means that it should print five decimal places, but it actually means lookup the number of decimal places to print from the argument number five (right? That's what I at least infer from my reverse engineering in those examples I wrote.)

**Comment by pnkfelix** [comment]
So yes, on further reflection, while these changes are good, I think one more thing is needed: An expansion of the text at http://doc.rust-lang.org/nightly/std/fmt/index.html#precision that explains how the actual precision argument as written is interpreted.

And then you should probably consider moving my examples that you have transcribed *into* that section on the `precision` formatting parameter.

**Comment by Manishearth** [comment]
I think an explicit mention of the differences between  `*` and `$`, and some more examples, would be helpful.

**Comment by pnkfelix** [comment]
So here is a draft of what we might write in the new `precision` section:

> ## Precision
> ...
> For floating-point types, this indicates how many digits after the decimal point should be printed.
> 
> There are three possible ways to specify the desired `precision`: an integer `.N`, an integer followed by dollar sign `.N$`, or an asterisk `.*`.
> * The first specification, `.N`, means the integer `N` itself is the precision.
> * The second, `.N$`, means use format *argument* `N` (which must be a `usize`) as the precision.
> * Finally,  `.*` means that the precision is held in the format argument that precedes the value that this `{...}` is printing.
> 
> For example, these: 
> ...

**Comment by pnkfelix** [comment]
(ah good doc is so hard...)

**Comment by Manishearth** [comment]
:+1:

File: src/libcollections/fmt.rs (hunk 2 above)
Comment (steveklabnik) [inline src/libcollections/fmt.rs]: :+1:

**Comment by steveklabnik** [comment]
> (ah good doc is so hard...)

Right? Thanks so much :)

**Comment by steveklabnik** [comment]
(Since @pnkfelix ended up writing the doc, i'll r+ it)

**Comment by steveklabnik** [comment]
Sneaking this into https://github.com/rust-lang/rust/pull/24786 as well

**Pull request merged**